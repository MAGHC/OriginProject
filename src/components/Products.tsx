import { useQuery } from '@tanstack/react-query';

import { useState } from 'react';

import { getProducts } from '../api/firebase.js';
import ProductCard from './ProductCard';
import ProductSkeleton from './skeleton/ProductSkeleton';

import { GetProductTypeI } from './../type/GetProductType';

const PRODUCT_WRAPPER = ' md:mx-[16rem]  grid grid-cols-1 sm:grid-cols-2 md:grid-cols-4 gap-8';

const MockSkeletonState = new Array(12);
MockSkeletonState.fill(null);

const CATEGORYIES = ['Women', 'Goods', 'Life', 'Special'];

const Products = () => {
  const [filter, setFilter] = useState<string | boolean>(false);

  const {
    error,
    isLoading,
    data: products,
  } = useQuery<GetProductTypeI[]>(['products'], getProducts, {
    select: (products) =>
      products.filter((product) => {
        return filter ? product.category === filter : product;
      }),
  });

  return (
    <>
      <ul className="sm:my-20 mx-[8rem] text-xl md:mx-[16rem] font-body text-highLight mt-[6.85rem] mb-[1.9rem] flex gap-12">
        {CATEGORYIES.map((category) => (
          <li className=" hover:scale-110 transition-all duration-200">
            <button onClick={() => setFilter(category)}>{category}</button>
          </li>
        ))}
      </ul>
      {isLoading && (
        <ul className={`${PRODUCT_WRAPPER}`}>
          {MockSkeletonState.map(() => (
            <ProductSkeleton />
          ))}
        </ul>
      )}
      {error && <p>에러</p>}
      {products && (
        <ul className={`${PRODUCT_WRAPPER}`}>
          {products.map((product) => (
            <ProductCard key={product.id} product={product} />
          ))}
        </ul>
      )}
    </>
  );
};

export default Products;
