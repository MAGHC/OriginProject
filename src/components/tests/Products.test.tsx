import { render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Products from './../Products';

import { withAllContext, withRouter } from '../../tests/utils';
import { Route } from 'react-router-dom';
import { mockProducts } from '../../tests/data';

describe('Products', () => {
  const mockAuth = {
    login: jest.fn(),
    logout: jest.fn(),
    uid: null,
    user: null,
  };

  const mockGetProdut = jest.fn();

  afterEach(() => {
    mockGetProdut.mockReset();
  });

  it('snapshot', () => {
    mockGetProdut.mockImplementation(() => mockProducts);

    const component = renderer
      .create(
        withAllContext(
          withRouter(
            <Route path="/" element={<Products getProducts={mockGetProdut}></Products>}></Route>,
          ),
          mockAuth,
        ),
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  it('error renders', async () => {
    mockGetProdut.mockImplementation(() => null);

    renderProduts();
    await waitFor(() => screen.findByText('에러'));
  });

  it('success renders', async () => {
    mockGetProdut.mockImplementation(() => mockProducts);
    renderProduts();

    await (waitFor(() => screen.findByText('test14')), { timeout: 3000 });
  });

  function renderProduts() {
    return render(
      withAllContext(
        withRouter(
          <Route path="/" element={<Products getProducts={mockGetProdut}></Products>}></Route>,
        ),
        mockAuth,
      ),
    );
  }
});

export {};
