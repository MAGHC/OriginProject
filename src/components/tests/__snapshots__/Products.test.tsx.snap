// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products snapshot 1`] = `
Array [
  <ul
    className="sm:my-20 mx-[8rem] text-xl md:mx-[16rem] font-body text-highLight mt-[6.85rem] mb-[1.9rem] flex gap-12"
  >
    <li
      className=" hover:scale-110 transition-all duration-200"
    >
      <button
        onClick={[Function]}
      >
        Women
      </button>
    </li>
    <li
      className=" hover:scale-110 transition-all duration-200"
    >
      <button
        onClick={[Function]}
      >
        Goods
      </button>
    </li>
    <li
      className=" hover:scale-110 transition-all duration-200"
    >
      <button
        onClick={[Function]}
      >
        Life
      </button>
    </li>
    <li
      className=" hover:scale-110 transition-all duration-200"
    >
      <button
        onClick={[Function]}
      >
        Special
      </button>
    </li>
  </ul>,
  <ul
    className=" md:mx-[16rem]  grid grid-cols-1 sm:grid-cols-2 md:grid-cols-4 gap-8"
  >
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
    <li
      className=" animate-pulse mx-auto "
    >
      <div
        className=" rounded-lg bg-slate-400 w-56 h-44"
      />
      <div
        className=" w-56 my-2 gap-4 flex justify-between"
      >
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
        <div
          className=" bg-slate-400 w-1/4 h-3"
        />
      </div>
    </li>
  </ul>,
]
`;
